examples/eSealing/src/lib.rs

```diff
@@ -0,0 +1,328 @@
+//!   and witness (sender_account) from the smart contract to prove that the
+//!   witness (sender_account) was in possession of that file at that time.
+//!
+//! Only accounts can register a file hash. During the registration, the
+//! timestamp is recorded together with the witness (sender_account) that
+//! registered the file hash. Each file hash can only be registered once.
+//! This is because we want to record the first time when the witness
+//! (sender_account) had access to that file. Re-registering a file hash (by a
+//! different witness) would not prove that the second witness is also in
+//! possession of that file because the second witness could have read the
+//! file hash during the initial registration transaction from the blockchain.
+use concordium_std::*;
+
+/// The different errors the contract can produce.
+#[derive(Serialize, Debug, PartialEq, Eq, Reject, SchemaType)]
+enum ContractError {
+    /// Failed parsing the parameter.
+    #[from(ParseError)]
+    ParseParams,
+    /// Failed logging: Log is full.
+    LogFull,
+    /// Failed logging: Log is malformed.
+    LogMalformed,
+    /// Only accounts can register a file hash.
+    OnlyAccount,
+    /// Each file hash can only be registered once.
+    AlreadyRegistered,
+}
+
+/// Mapping the logging errors to ContractError.
+impl From<LogError> for ContractError {
+    fn from(le: LogError) -> Self {
+        match le {
+            LogError::Full => Self::LogFull,
+            LogError::Malformed => Self::LogMalformed,
+        }
+    }
+}
+
+/// The state tracked for each file.
+#[derive(Serial, Deserial, Clone, Copy, SchemaType)]
+struct FileState {
+    /// The timestamp when this file hash was registered.
+    timestamp: Timestamp,
+    /// The witness (sender_account) that registered this file hash.
+    witness:   AccountAddress,
+}
+
+/// FileHash type is a fixed-length byte array.
+type FileHash = [u8; 32];
+
+/// The contract state.
+#[derive(Serial, DeserialWithState, StateClone)]
+#[concordium(state_parameter = "S")]
+struct State<S> {
+    files: StateMap<FileHash, FileState, S>,
+}
+
+impl<S: HasStateApi> State<S> {
+    /// Create a new state with no files registered.
+    fn new(state_builder: &mut StateBuilder<S>) -> Self {
+        State {
+            files: state_builder.new_map(),
+        }
+    }
+
+    /// Check if a file exists.
+    fn file_exists(&self, file_hash: &FileHash) -> bool {
+        let file = self.files.get(file_hash);
+
+        file.is_some()
+    }
+
+    /// Get recorded FileState (timestamp and witness) from a specific file
+    /// hash.
+    fn get_file_state(&self, file_hash: FileHash) -> ReceiveResult<Option<FileState>> {
+        Ok(self.files.get(&file_hash).map(|v| *v))
+    }
+
+    /// Add a new file hash.
+    fn add_file(&mut self, file_hash: FileHash, timestamp: Timestamp, witness: AccountAddress) {
+        self.files.insert(file_hash, FileState {
```

Comment:
Minor improvement:
[`StateMap.insert`](https://docs.rs/concordium-std/latest/concordium_std/struct.StateMap.html#method.insert) returns whether the map already contained the key. If you make `add_file` return an error in the case of file already existing, you don't need to first look it up in the `registerFile` receive function.
Might be less readable, so I am okay without the improvement.